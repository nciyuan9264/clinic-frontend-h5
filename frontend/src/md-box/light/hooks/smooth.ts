import { useEffect, useRef, useState } from 'react';

import { defaults, isBoolean, isNumber, isString } from 'lodash-es';
import { useInterval } from 'ahooks';

import {
  useEstablished,
  useForceUpdate,
  useLatestFunction,
  usePreviousRef,
} from './common';

const INITIAL_SPEED = 20 / 1000;

const EXPECTED_BUFFER_SIZE = 15;

const ELASTIC_COEFFICIENT = 1 / 700 / 1000;

const TICK_DURATION = (1000 / 60) * 2;

const findCommonPrefixLength = (first: string, second: string) => {
  let leftIndex = 0;
  let rightIndex = Math.max(first.length, second.length);

  while (leftIndex !== rightIndex && leftIndex !== rightIndex - 1) {
    const midIndex = Math.floor((leftIndex + rightIndex) / 2);

    // eslint-disable-next-line @typescript-eslint/prefer-string-starts-ends-with
    if (first.slice(0, midIndex) === second.slice(0, midIndex)) {
      leftIndex = midIndex;
    } else {
      rightIndex = midIndex;
    }
  }

  return leftIndex;
};

export interface SmoothOnUpdateParams {
  /**
   * 当前输出的文本
   */
  text: string;
  /**
   * 当前输出速度，单位：字 / 秒
   */
  speed: number;
}

export interface SmoothOnTextBreakParams {
  /**
   * 破坏性更新之前的文本
   */
  prevText: string;
  /**
   * 破坏性更新之后的文本
   */
  currentText: string;
}

export interface SmoothOptions {
  /**
   * 是否开启流式输出平滑化
   * @default true
   */
  enable?: boolean;
  /**
   * 当传入此值时，启用首屏文本直出（不开启平滑）的功能，此值代表当首屏文本不大于多少时从1字开始
   * 平滑输出，否则以该文本为起始进度，第二次更新再继续平滑输出
   *
   * 不传代表任意时刻都启用平滑化
   *
   * 例如，当传入值为5时：
   * - 首次传入文本为abc，则abc从空平滑化输出
   * - 首次传入文本为abcdef，则abcdef首屏直接展示，不平滑化输出
   *
   * 本功能应用场景为：当流式输出一半时切出，假如不开启此功能，再切回会从空串重新开始平滑，开启时
   * 能自动以超过指定长度的首文本作为起始进度
   * @default Infinity
   */
  maxFirstTextSmoothSize?: number;
  /**
   * 启动平滑化尾优化。开启时，当enable从true变为false时，会以更快的速度平滑化输出到文本结束，关闭时当enable变为false时，则会直接输出全文
   * 默认为false
   * @default false
   */
  trailingSmooth?: boolean;
  /**
   * 暂停流式动画
   * @default false
   */
  pause?: boolean;
  /**
   * 开启平滑化时的实际渲染完成回调，用于获取真实结束时间
   */
  onSmoothFinished?: () => void;
  /**
   * 平滑后的内容更新时回调
   */
  onUpdate?: (params: SmoothOnUpdateParams) => void;
  /**
   * 文本破坏性更新回调，破坏性更新指前面的文本非最新文本的前缀，因此无法再继续平滑化，上屏内容
   * 需要立即重置为最新的文本
   */
  onTextBreak?: (params: SmoothOnTextBreakParams) => void;
  /**
   * 流式平滑最大的速度，单位：字符 / 秒
   * @default Infinity
   */
  maxSpeed?: number;
  /**
   * 弹簧平衡位置（即预期剩余缓冲区长度），单位：字符
   * @default 15
   */
  expectedBufferSize?: number;
  /**
   * 系数（弹性系数/质量），数值越大，则跟随内容越及时、平顺性越低
   * @default 1/700/1000
   */
  elasticCoefficient?: number;
  /**
   * 最小更新时间间隔，数值越小，动画播放越快，单位：毫秒
   * @default 100/3
   */
  tickDuration?: number;
  /**
   * 初始速度，单位：字符 / 秒
   * @default 20/1000
   */
  initialSpeed?: number;
  /**
   * 当出现文本破坏式更新时（非拼接式更新），如何重置光标位置
   * - prefix: 重置到公共前缀
   * - end: 重置到文本末尾
   * - start: 重置到文本开头
   * @default 'prefix'
   */
  breakMode?: 'prefix' | 'end' | 'start';
}

export interface UseSmoothTextReturnValue {
  text: string;
  /** 将游标设置到当前文本末尾处 */
  flushCursor: () => void;
}

// eslint-disable-next-line max-lines-per-function
export const useSmoothText = (
  text: string,
  smoothConfig: boolean | SmoothOptions = {},
): UseSmoothTextReturnValue => {
  const {
    maxFirstTextSmoothSize,
    trailingSmooth,
    enable,
    pause,
    onSmoothFinished,
    onUpdate,
    onTextBreak,
    maxSpeed: _maxSpeed,
    expectedBufferSize,
    elasticCoefficient,
    tickDuration,
    initialSpeed,
    breakMode,
  } = defaults(isBoolean(smoothConfig) ? {} : { ...smoothConfig }, {
    maxFirstTextSmoothSize: Infinity,
    trailingSmooth: false,
    pause: false,
    enable: Boolean(smoothConfig),
    expectedBufferSize: EXPECTED_BUFFER_SIZE,
    elasticCoefficient: ELASTIC_COEFFICIENT,
    tickDuration: TICK_DURATION,
    initialSpeed: INITIAL_SPEED,
    breakMode: 'prefix',
  });

  const maxSpeed = isNumber(_maxSpeed) && _maxSpeed > 0 ? _maxSpeed : Infinity;

  const prevTextRef = usePreviousRef(text);

  const currentPosition = useRef(
    text.length <= maxFirstTextSmoothSize ? 1 : text.length,
  );

  const indexRef = useRef(currentPosition.current);

  const hasEnable = useEstablished(enable);

  // 最开始为 false 的时候不要流式，只有 enable 从 true 变为 false 的时候才要结尾优化一下
  const running =
    enable ||
    (hasEnable && trailingSmooth && currentPosition.current < text.length);

  const [isAtEnding, setIsAtEnding] = useState(false);

  const currentSpeed = useRef(initialSpeed);

  const isAtBreak = () => {
    return (
      isString(prevTextRef.current) && !text.startsWith(prevTextRef.current)
    );
  };

  /** 如下代码要实现发现文本破坏式更新时，同步地更新下标，使得内容和下标更新同步 */
  if (isAtBreak()) {
    let initialPosition = 0;

    if (breakMode === 'start') {
      initialPosition = 0;
    } else if (breakMode === 'end') {
      initialPosition = Math.max(text.length - 1, 0);
    } else if (breakMode === 'prefix') {
      initialPosition = Math.max(
        1,
        findCommonPrefixLength(prevTextRef.current ?? '', text),
      );
    }

    currentPosition.current = initialPosition;

    indexRef.current = Math.floor(currentPosition.current);
  }

  const index = indexRef.current;

  const currentText = running && !isAtEnding ? text.slice(0, index) : text;

  const handleForceUpdate = useForceUpdate();

  const setIndex = (newIndex: number) => {
    if (newIndex === indexRef.current) {
      return;
    }

    indexRef.current = newIndex;

    handleForceUpdate();
  };

  const handleFlushCursor = useLatestFunction(() => {
    setIsAtEnding(true);
  });

  useEffect(() => {
    if (isAtEnding) {
      setIndex(text.length);

      setIsAtEnding(false);

      currentPosition.current = text.length;
    }
  }, [isAtEnding]);

  useEffect(() => {
    if (!running) {
      onSmoothFinished?.();
    }
  }, [running]);

  useInterval(
    () => {
      if (!isAtBreak()) {
        const endPosition = Math.max(
          text.length +
            (!enable && trailingSmooth ? 2 * expectedBufferSize : 0),
          expectedBufferSize * 2,
        );

        const restSpace = endPosition - currentPosition.current;

        const balanceDistance = restSpace - expectedBufferSize;

        const acceleration = elasticCoefficient * balanceDistance;

        currentSpeed.current = Math.min(
          Math.max(0, currentSpeed.current + tickDuration * acceleration),
          /** 从“字/每秒”换算为“字/毫秒” */
          maxSpeed / 1000,
        );

        const diffPosition = tickDuration * currentSpeed.current;

        currentPosition.current = Math.min(
          text.length,
          currentPosition.current + diffPosition,
        );

        setIndex(Math.floor(currentPosition.current));
      } else {
        onTextBreak?.({
          prevText: prevTextRef.current ?? '',
          currentText: text,
        });

        currentSpeed.current = initialSpeed;

        prevTextRef.current = undefined;
      }
    },
    running && !pause ? tickDuration : undefined,
  );

  useEffect(() => {
    onUpdate?.({
      text: currentText,
      speed: currentSpeed.current * 1000,
    });
  }, [index]);

  return {
    text: currentText,
    flushCursor: handleFlushCursor,
  };
};
