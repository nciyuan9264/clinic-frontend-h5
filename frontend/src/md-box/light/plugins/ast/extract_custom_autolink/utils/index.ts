export { HyperNodeType, getHyperLinkNodes } from './parser';

export const completeProtocol = (raw: string) => {
  if (raw.match(/https?:\/\//)) {
    return raw;
  }
  return `https://${raw}`;
};
