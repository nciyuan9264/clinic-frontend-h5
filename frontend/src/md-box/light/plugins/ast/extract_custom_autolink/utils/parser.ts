import emailValidator from 'email-validator';

import { urlRegExp } from './regex';

export enum HyperNodeType {
  text,
  link,
}

export type HyperNode =
  | {
      type: HyperNodeType.text;
      text: string;
    }
  | {
      type: HyperNodeType.link;
      url: string;
    };

export type HyperNodeTitled = HyperNode & {
  title: string;
};

const openAndCloseCharPairs: Record<string, string> = {
  '(': ')',
  '[': ']',
  '{': '}',
  '<': '>',
};
const openChars = Object.keys(openAndCloseCharPairs);
const closeChars = Object.values(openAndCloseCharPairs);
const unreasonableSuffix = ['.', ',', '!', '?', "'", ':', ';', '@', '#'];

/**
 * 分割合理的URL
 * @returns 返回字符串中合理的URL的长度，范围[0, input.length]
 */
const getAReasonableUrl = (input: string) => {
  let index = -1;
  const stack: Array<{
    char: string;
    index: number;
  }> = [];
  for (let i = 0; i < input.length; ++i) {
    const char = input[i];
    if (openChars.includes(char)) {
      stack.push({
        char,
        index: i,
      });
    } else if (closeChars.includes(char)) {
      if (!stack.length) {
        index = i;
        break;
      } else {
        const { char: lastOpenChar } = stack[stack.length - 1];
        if (char === openAndCloseCharPairs[lastOpenChar]) {
          stack.pop();
        } else {
          ({ index } = stack[0]);
          break;
        }
      }
    }
  }
  if (stack.length) {
    ({ index } = stack[0]);
  }
  if (index === -1) {
    index = input.length;
  }

  while (index > 1 && unreasonableSuffix.includes(input[index - 1])) {
    index--;
  }
  return input.slice(0, index);
};

/**
 * 分割字符串中的超链接和常规字符，输出格式化数据 HyperNode[]
 */
export const getHyperLinkNodes = (str: string): HyperNode[] => {
  const nodes: HyperNode[] = [];
  let cur = 0;
  urlRegExp.lastIndex = 0;
  while (true) {
    const result = urlRegExp.exec(str);
    if (!result) {
      break;
    }

    const { index, 0: content } = result;
    const text = str.slice(cur, index);

    // ① 文字部分
    if (text) {
      nodes.push({
        type: HyperNodeType.text,
        text,
      });
    }

    // ② 链接/markdown/email 部分，email 被渲染为文字，code block / inline code 中的链接也当做文字
    if (content) {
      /** 1. 非正常前缀 */
      const abnormalPrefixMatchResult = /^(@|#|\+|-|%)+/.exec(content);
      if (abnormalPrefixMatchResult) {
        nodes.push({
          type: HyperNodeType.text,
          text: abnormalPrefixMatchResult[0],
        });
        cur = result.index + abnormalPrefixMatchResult[0].length;
        urlRegExp.lastIndex = cur;
      } else {
        /** 2. 合理Url作为前缀 */
        const subUrl = getAReasonableUrl(content);

        if (subUrl) {
          if (emailValidator.validate(subUrl)) {
            nodes.push({
              type: HyperNodeType.text,
              text: subUrl,
            });
          } else {
            nodes.push({
              type: HyperNodeType.link,
              url: subUrl,
            });
          }

          cur = result.index + subUrl.length;
          urlRegExp.lastIndex = cur;
        } else {
          const skipStr = content.slice(0, 1);

          nodes.push({
            type: HyperNodeType.text,
            text: skipStr,
          });

          cur = result.index + skipStr.length;
          urlRegExp.lastIndex = cur;
        }
      }
    }
  }
  if (cur !== str.length) {
    nodes.push({ type: HyperNodeType.text, text: str.slice(cur) });
  }
  return nodes;
};
