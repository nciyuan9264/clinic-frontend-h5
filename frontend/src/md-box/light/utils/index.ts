export * from './micromark';
export * from './marked';
export * from './renderer';
export * from './common';
export * from './hop';
export * from './plugin';
