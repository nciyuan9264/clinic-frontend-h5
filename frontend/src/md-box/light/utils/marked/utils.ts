/** 源自 markedjs 源码 */

const escapeTest = /[&<>"']/;
const escapeReplace = new RegExp(escapeTest.source, 'g');
const escapeTestNoEncode = /[<>"']|&(?!(#\d{1,7}|#[Xx][a-fA-F0-9]{1,6}|\w+);)/;
const escapeReplaceNoEncode = new RegExp(escapeTestNoEncode.source, 'g');

const escapeReplacements: Record<string, string> = {
  '&': '&amp;',
  '<': '&lt;',
  '>': '&gt;',
  '"': '&quot;',
  "'": '&#39;',
};

export const escapeMarkdownText = (html: string, encode = false) => {
  const getEscapeReplacement = (ch: string) => escapeReplacements[ch];

  if (encode) {
    if (escapeTest.test(html)) {
      return html.replace(escapeReplace, getEscapeReplacement);
    }
  } else {
    if (escapeTestNoEncode.test(html)) {
      return html.replace(escapeReplaceNoEncode, getEscapeReplacement);
    }
  }
  return html;
};
