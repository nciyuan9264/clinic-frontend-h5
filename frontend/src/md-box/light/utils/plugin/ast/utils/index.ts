export * from './update_ast';
