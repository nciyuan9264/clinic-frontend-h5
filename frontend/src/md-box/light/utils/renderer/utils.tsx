import React, { createElement } from 'react';

import { isRegExp } from 'lodash-es';
import { DOMNode, Element, Text, domToReact } from 'html-react-parser';

import { RenderFunction } from '../../type';

export const isTextType = (node: DOMNode): node is Text => {
  return node.type === 'text';
};

export const isEndlineText = (node: DOMNode): node is Text => {
  return isTextType(node) && /^\n+$/.test(getInnerText(node));
};

/**
 * 此处断言的是Element的子类型：即为A标签的类型，但是Element不可进一步进行类型收窄，因此
 * 返回node is Element
 */
export const isElementType = (
  node: DOMNode,
  ...names: (string | RegExp)[]
): node is Element => {
  if (!('attribs' in node) || typeof node.attribs !== 'object') {
    return false;
  }
  return (
    node.type === 'tag' &&
    (!names?.length ||
      names.some((item) => {
        if (isRegExp(item)) {
          return item.test(node.name);
        } else {
          return item === node.name;
        }
      }))
  );
};

export const getInnerText = (node: DOMNode): string => {
  if (isTextType(node)) {
    return node.data;
  }
  if (isElementType(node)) {
    return node.children.map(getInnerText).join('');
  }
  return '';
};

/**
 * 检测字符或字符串是否包含中文字符
 * https://github.com/anton-bot/contains-chinese/blob/master/index.js#L7
 * @param char 字符或字符串
 * @returns 是否包含中文字符
 */
export const detectIsChinese = (char: string) => {
  const HAN_REGEX =
    // eslint-disable-next-line max-len
    /[\u2E80-\u2E99\u2E9B-\u2EF3\u2F00-\u2FD5\u3005\u3007\u3021-\u3029\u3038-\u303B\u3400-\u4DB5\u4E00-\u9FD5\uF900-\uFA6D\uFA70-\uFAD9]/;
  return HAN_REGEX.test(char);
};

export const renderDom: RenderFunction<false> = (node) => {
  return <>{domToReact([node])}</>;
};

export const renderReactElement = <P extends Record<string, unknown>>(
  tag: string,
  props: P,
  children?: React.ReactNode[],
) => {
  if (props.className === 'markdown-root') {
    return <>{children}</>;
  } else {
    return (
      <>{createElement(tag, props, children?.length ? children : undefined)}</>
    );
  }
};
