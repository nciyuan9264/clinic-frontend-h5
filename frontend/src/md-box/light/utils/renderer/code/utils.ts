import { DOMNode, Element } from 'html-react-parser';

import { isElementType } from '../utils';

const DEFAULT_LANGUAGE = 'plaintext';

const getCodeElement = (node: DOMNode): Element | undefined => {
  if (isElementType(node, 'code')) {
    return node;
  }

  if (isElementType(node, 'pre')) {
    return node.childNodes.find((childNode): childNode is Element =>
      isElementType(childNode, 'code'),
    );
  }
};

export const getLanguageFromElement = (
  node: DOMNode,
  defaultLanguage = DEFAULT_LANGUAGE,
): string => {
  const codeElement = getCodeElement(node);

  if (codeElement) {
    return (
      codeElement.attribs?.class?.match(/language-(.*)/)?.[1] || defaultLanguage
    );
  }

  return defaultLanguage;
};

export const getMetaFromElement = (node: DOMNode, removeBuiltIn = true) => {
  const codeElement = getCodeElement(node);

  const metaString = codeElement?.attribs['data-meta'];

  if (!metaString) {
    return;
  }

  if (!removeBuiltIn) {
    return metaString;
  }

  return (
    metaString
      .split(/\s+/)
      .filter((item) => !item.match(/^__[^_].*/))
      .join(' ') || undefined
  );
};

export const getIndentFromCodeElememt = (node: DOMNode) => {
  const metaString = getMetaFromElement(node, false);

  if (!metaString) {
    return 0;
  }

  const matchResult = metaString.match(
    /(\s|^)__indent=(?<indent>[0-9]+)(\s|$)/,
  );

  if (!matchResult) {
    return 0;
  }

  const indentValue = Number(matchResult.groups?.indent);

  return isNaN(indentValue) ? 0 : indentValue;
};
