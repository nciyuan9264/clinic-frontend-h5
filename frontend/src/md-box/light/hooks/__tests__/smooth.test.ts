import { concat } from 'lodash-es';
import { act, renderHook } from '@testing-library/react';

import { renderHookInterval } from './utils';
import { useSmoothText } from '../smooth';

const smoothedText = '123456789012345678901234567890';

describe('平滑化方法测试（useSmoothText）', () => {
  afterEach(() => {
    jest.useRealTimers();
  });

  beforeEach(() => {
    jest.useFakeTimers();
  });

  test('初始固定文本，验证从头到尾平滑化', async () => {
    const renderHandle = await renderHookInterval({
      useHook: () => useSmoothText(smoothedText),
      mapper: (item) => item.text,
      interval: 30,
      loop: 50,
    });

    expect(renderHandle.results).toMatchSnapshot('正常流式平滑化直到结束');
  });

  test('调用 flushCursor，验证光标直接调整到末尾', async () => {
    const renderHandle = await renderHookInterval({
      useHook: () => useSmoothText(smoothedText),
      mapper: (item) => item.text,
      interval: 30,
      loop: 10,
    });

    await act(() => {
      renderHandle.result.current.flushCursor();
    });

    const secondRenderHandle = await renderHookInterval({
      useHook: renderHandle,
      mapper: (item) => item.text,
      interval: 30,
      loop: 10,
    });

    expect(
      concat(renderHandle.results, secondRenderHandle.results),
    ).toMatchSnapshot('第十次后光标调整到末尾');
  });

  test('在启用平滑化的同时调用 flushCursor，验证初始从改位置开始', async () => {
    let currentText = smoothedText;

    let currentEnable = false;

    const renderHandle = await act(() =>
      renderHook(() => useSmoothText(currentText, currentEnable)),
    );

    await act(() => {
      currentEnable = true;

      renderHandle.result.current.flushCursor();

      renderHandle.rerender();
    });

    currentText += smoothedText;

    const { results } = await renderHookInterval({
      useHook: renderHandle,
      mapper: (item) => item.text,
      interval: 30,
      loop: 30,
    });

    expect(results).toMatchSnapshot('直接从初始文本末尾作为开始进行流式');
  });
});
