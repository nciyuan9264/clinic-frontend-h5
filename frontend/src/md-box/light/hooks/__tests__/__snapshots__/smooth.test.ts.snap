// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`平滑化方法测试（useSmoothText） 初始固定文本，验证从头到尾平滑化: 正常流式平滑化直到结束 1`] = `
[
  "1",
  "1",
  "1",
  "12",
  "123",
  "123",
  "1234",
  "12345",
  "123456",
  "1234567",
  "1234567",
  "12345678",
  "12345678",
  "123456789",
  "1234567890",
  "12345678901",
  "123456789012",
  "1234567890123",
  "1234567890123",
  "12345678901234",
  "123456789012345",
  "1234567890123456",
  "12345678901234567",
  "12345678901234567",
  "123456789012345678",
  "1234567890123456789",
  "1234567890123456789",
  "12345678901234567890",
  "123456789012345678901",
  "1234567890123456789012",
  "12345678901234567890123",
  "123456789012345678901234",
  "123456789012345678901234",
  "1234567890123456789012345",
  "1234567890123456789012345",
  "12345678901234567890123456",
  "123456789012345678901234567",
  "123456789012345678901234567",
  "1234567890123456789012345678",
  "12345678901234567890123456789",
  "12345678901234567890123456789",
  "123456789012345678901234567890",
  "123456789012345678901234567890",
  "123456789012345678901234567890",
  "123456789012345678901234567890",
  "123456789012345678901234567890",
  "123456789012345678901234567890",
  "123456789012345678901234567890",
  "123456789012345678901234567890",
  "123456789012345678901234567890",
  "123456789012345678901234567890",
]
`;

exports[`平滑化方法测试（useSmoothText） 在启用平滑化的同时调用 flushCursor，验证初始从改位置开始: 直接从初始文本末尾作为开始进行流式 1`] = `
[
  "123456789012345678901234567890",
  "123456789012345678901234567890",
  "123456789012345678901234567890",
  "1234567890123456789012345678901",
  "12345678901234567890123456789012",
  "12345678901234567890123456789012",
  "123456789012345678901234567890123",
  "1234567890123456789012345678901234",
  "12345678901234567890123456789012345",
  "123456789012345678901234567890123456",
  "123456789012345678901234567890123456",
  "1234567890123456789012345678901234567",
  "1234567890123456789012345678901234567",
  "12345678901234567890123456789012345678",
  "123456789012345678901234567890123456789",
  "1234567890123456789012345678901234567890",
  "12345678901234567890123456789012345678901",
  "123456789012345678901234567890123456789012",
  "1234567890123456789012345678901234567890123",
  "1234567890123456789012345678901234567890123",
  "12345678901234567890123456789012345678901234",
  "123456789012345678901234567890123456789012345",
  "1234567890123456789012345678901234567890123456",
  "1234567890123456789012345678901234567890123456",
  "12345678901234567890123456789012345678901234567",
  "123456789012345678901234567890123456789012345678",
  "1234567890123456789012345678901234567890123456789",
  "12345678901234567890123456789012345678901234567890",
  "123456789012345678901234567890123456789012345678901",
  "1234567890123456789012345678901234567890123456789012",
  "1234567890123456789012345678901234567890123456789012",
]
`;

exports[`平滑化方法测试（useSmoothText） 调用 flushCursor，验证光标直接调整到末尾: 第十次后光标调整到末尾 1`] = `
[
  "1",
  "1",
  "1",
  "12",
  "123",
  "123",
  "1234",
  "12345",
  "123456",
  "1234567",
  "1234567",
  "123456789012345678901234567890",
  "123456789012345678901234567890",
  "123456789012345678901234567890",
  "123456789012345678901234567890",
  "123456789012345678901234567890",
  "123456789012345678901234567890",
  "123456789012345678901234567890",
  "123456789012345678901234567890",
  "123456789012345678901234567890",
  "123456789012345678901234567890",
  "123456789012345678901234567890",
]
`;
